packages/apps/shopify-app-remix/src/server/authenticate/admin/strategies/__tests__/merchant-custom/admin-client.test.ts

```diff
@@ -0,0 +1,132 @@
+import {restResources} from '@shopify/shopify-api/rest/admin/2023-04';
+import {ApiVersion, LATEST_API_VERSION, Session} from '@shopify/shopify-api';
+
+import {AppDistribution} from '../../../../../types';
+import {AdminApiContext} from '../../../../../clients';
+import {shopifyApp} from '../../../../..';
+import {
+  APP_URL,
+  TEST_SHOP,
+  expectAdminApiClient,
+  getThrownResponse,
+  mockExternalRequest,
+  setupValidCustomAppSession,
+  testConfig,
+} from '../../../../../__test-helpers';
+
+describe('admin.authenticate context', () => {
+  expectAdminApiClient(async () => {
+    const {
+      admin,
+      expectedSession,
+      session: actualSession,
+    } = await setUpMerchantCustomFlow();
+
+    return {admin, expectedSession, actualSession};
+  });
+  describe.each([
+    {
+      testGroup: 'REST client',
+      mockRequest: mockRestRequest,
+      action: async (admin: AdminApiContext, _session: Session) =>
+        admin.rest.get({path: '/customers.json'}),
+    },
+    {
+      testGroup: 'REST resources',
+      mockRequest: mockRestRequest,
+      action: async (admin: AdminApiContext, session: Session) =>
+        admin.rest.resources.Customer.all({session}),
+    },
+    {
+      testGroup: 'GraphQL client',
+      mockRequest: mockGraphqlRequest(),
+      action: async (admin: AdminApiContext, _session: Session) =>
+        admin.graphql('{ shop { name } }'),
+    },
+    {
+      testGroup: 'GraphQL client with options',
+      mockRequest: mockGraphqlRequest('2021-01' as ApiVersion),
+      action: async (admin: AdminApiContext, _session: Session) =>
+        admin.graphql(
+          'mutation myMutation($ID: String!) { shop(ID: $ID) { name } }',
+          {
+            variables: {ID: '123'},
+            apiVersion: '2021-01' as ApiVersion,
+            headers: {custom: 'header'},
+            tries: 2,
+          },
+        ),
+    },
+  ])(
+    '$testGroup re-authentication',
+    ({testGroup: _testGroup, mockRequest, action}) => {
+      it('returns 401 when receives a 401 response on fetch requests', async () => {
```

Comment:
I agree with this approach, there is no way we can automatically recover from this scenario as the token would have to have changed in the admin side.